cosmos/scheduler/cmd/scheduler.go

```diff
@@ -0,0 +1,266 @@
+package scheduler
+
+import (
+	"context"
+	"log"
+
+	"github.com/mongodb/mongo-go-driver/bson"
+	"github.com/mongodb/mongo-go-driver/mongo"
+	"github.com/scality/zenko/cosmos/api/types/v1alpha1"
+	clientV1alpha1 "github.com/scality/zenko/cosmos/clientset/v1alpha1"
+	"github.com/scality/zenko/cosmos/scheduler/pkg"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// Scheduler represents a Cosmos Scheduler instance
+type Scheduler struct {
+	KubeAlpha         *clientV1alpha1.ExampleV1Alpha1Client
+	KubeClientset     *kubernetes.Clientset
+	Pensieve          *pensieve.Helper
+	Namespace         string
+	Cloudserver       string
+	StorageClass      string
+	MongodbClient     *mongo.Client
+	SecretName        string
+	IngestionSchedule string
+}
+
+// Run starts the Cosmos Scheduler
+func (s *Scheduler) Run() {
+	log.Println("starting cosmos scheduler")
+	err := s.configureIngestionSecret()
+	if err != nil {
+		log.Panicln(err.Error())
+	}
+	quit := make(chan bool)
+	overlayUpdates := s.watchOverlayUpdates()
+	for {
+		go s.run(quit)
+		<-overlayUpdates
+		log.Println("got an overlay update")
+		quit <- true
+	}
+}
+
+func (s *Scheduler) configureIngestionSecret() error {
+	secret, err := s.getIngestionSecret()
+	if err != nil {
+		return err
+	}
+	if secret != "" {
+		s.SecretName = secret
+		log.Println("ingestion credentials secret found")
+		return nil
+	}
+	log.Println("ingestion secret not found")
+	secret, err = s.createIngestionSecret()
+	if err != nil {
+		return err
+	}
+	s.SecretName = secret
+	log.Println("ingestion secret created successfully")
+	return nil
+}
+
+func (s *Scheduler) run(quit chan bool) error {
+	locationBson, err := s.getCosmosLocationBson()
+	if err != nil {
+		return err
+	}
+	if len(locationBson) == 0 {
+		return nil
+	}
+	collection := s.MongodbClient.Database("metadata").Collection("__metastore")
+	ctx := context.Background()
+	cur, err := collection.Watch(ctx, mongo.Pipeline{
+		{{"$match", bson.D{
+			{"$or", locationBson},
+			{"fullDocument.value.ingestion.status", "enabled"},
+		}}},
+		{{"$project", bson.D{
+			{"operationType", 1},
+			{"fullDocument.value.ownerDisplayName", 1},
+			{"fullDocument.value.name", 1},
+			{"fullDocument.value.locationConstraint", 1},
+			{"fullDocument.value.deleted", 1},
+		}}},
+	})
+	if err != nil {
+		log.Println(err)
+		return err
+	}
+	ch := make(chan BucketTransaction)
+	go func() {
+		defer cur.Close(ctx)
+		for {
+			if cur.Next(ctx) {
+				var elem BucketTransaction
+				if err := cur.Decode(&elem); err != nil {
+					log.Println(err)
+					return
+				}
+				ch <- elem
+			} else {
+				return
+			}
+			if err := cur.Err(); err != nil {
```

Comment:
should this be in the `else` statement? correct me if I am wrong, this case doesn't seem reachable if `cur.Next() == false`

Reply:
That is correct